src/Common/src/System/Security/Cryptography/ECDiffieHellmanDerivation.cs

```diff
@@ -0,0 +1,243 @@
+                Debug.Assert(secretAgreement == null);
+
+                hash.AppendData(secretAppend);
+
+                return hash.GetHashAndReset();
+            }
+        }
+
+        internal static byte[] DeriveKeyFromHmac(
+            ECDiffieHellmanPublicKey otherPartyPublicKey,
+            HashAlgorithmName hashAlgorithm,
+            byte[] hmacKey,
+            ReadOnlySpan<byte> secretPrepend,
+            ReadOnlySpan<byte> secretAppend,
+            DeriveSecretAgreement deriveSecretAgreement)
+        {
+            Debug.Assert(otherPartyPublicKey != null);
+            Debug.Assert(!string.IsNullOrEmpty(hashAlgorithm.Name));
+
+            // If an hmac key is provided then calculate
+            // HMAC(hmacKey, prepend || derived || append)
+            //
+            // Otherwise, calculate
+            // HMAC(derived, prepend || derived || append)
+
+            bool useSecretAsKey = hmacKey == null;
+            GCHandle pinHandle = default;
+
+            if (useSecretAsKey)
+            {
+                hmacKey = deriveSecretAgreement(otherPartyPublicKey, null);
+                Debug.Assert(hmacKey != null);
+                pinHandle = GCHandle.Alloc(hmacKey, GCHandleType.Pinned);
+            }
+
+            using (IncrementalHash hash = IncrementalHash.CreateHMAC(hashAlgorithm, hmacKey))
+            {
+                hash.AppendData(secretPrepend);
+
+                if (useSecretAsKey)
+                {
+                    hash.AppendData(hmacKey);
+                    Array.Clear(hmacKey, 0, hmacKey.Length);
+                    Debug.Assert(pinHandle.IsAllocated);
+                    pinHandle.Free();
+                }
+                else
+                {
+                    byte[] secretAgreement = deriveSecretAgreement(otherPartyPublicKey, hash);
+                    // We want the side effect, and it should not have returned the answer.
+                    Debug.Assert(secretAgreement == null);
+                }
+
+                hash.AppendData(secretAppend);
+                return hash.GetHashAndReset();
+            }
+        }
+
+        internal static byte[] DeriveKeyTls(
+            ECDiffieHellmanPublicKey otherPartyPublicKey,
+            ReadOnlySpan<byte> prfLabel,
+            ReadOnlySpan<byte> prfSeed,
+            DeriveSecretAgreement deriveSecretAgreement)
+        {
+            Debug.Assert(otherPartyPublicKey != null);
+
+            if (prfSeed.Length != 64)
+            {
+                throw new CryptographicException(SR.Cryptography_TlsRequires64ByteSeed);
+            }
+
+            // Windows produces a 48-byte output, so that's what we do, too.
+            byte[] ret = new byte[48];
+
+            const int Sha1Size = 20;
+            const int Md5Size = 16;
+
+            byte[] secretAgreement = deriveSecretAgreement(otherPartyPublicKey, null);
+            GCHandle handle = GCHandle.Alloc(secretAgreement, GCHandleType.Pinned);
```

Comment:
Why does this need to be pinned?  It looks like handle is never used; how about using fixed instead?